File: euler/p0049.py
#!/usr/local/bin/python

import bisect
import math
import sys

#from "/Users/khmacdonald/code/SVN/trunk/euler/utilities/" import primes
sys.path.append("/Users/khmacdonald/code/SVN/trunk/euler/utilities/")
from primes import sieve

av = sys.argv
ac = len(av)

perm4 = [
    [3,2,1,0],
    [3,2,0,1],
    [3,1,2,0],
    [3,1,0,2],
    [3,0,2,1],
    [3,0,1,2], # 6
    [2,3,1,0],
    [2,3,0,1],
    [2,1,3,0],
    [2,1,0,3],
    [2,0,3,1],
    [2,0,1,3], # 12
    [1,3,2,0],
    [1,3,0,2],
    [1,2,3,0],
    [1,2,0,3],
    [1,0,3,2],
    [1,0,2,3], # 18
    [0,3,2,1],
    [0,3,1,2],
    [0,2,3,1],
    [0,2,1,3],
    [0,1,3,2],
    [0,1,2,3]  # 24
]

def get_dig ( p ):
    d = []
    d.append(p%10)
    p = p / 10
    d.append(p%10)
    p = p / 10
    d.append(p%10)
    p = p / 10
    d.append(p%10)
    return d

def perms ( p, sv ):
    global perm4
    d = get_dig(p)
    cnt = 0
    pm = []
    for perm in perm4:
        w,x,y,z = perm
        ptest = d[w]*1000 + d[x]*100 + d[y]*10 + d[z]
        if ptest in sv.plist and ptest not in pm:
            pm.append(ptest)
    return pm

def get_dig_cnt ( n ):
    d = {}
    while n:
        dig = n%10
        n   = n/10
        if dig in d:
            d[dig] = d[dig]+1
        else:
            d[dig] = 1
    return d

def is_perm ( m, n ):
    dm = get_dig_cnt(m)
    dn = get_dig_cnt(n)
    for d in dm:
        if d not in dn:
            return False
        if dm[d]!=dn[d]:
            return False
    return True

def write_primes ( sv ):
    fname = "p0049_prime_list.txt"
    fd = open(fname,"w")

    for k in range(len(sv.plist)):
        fd.write("%6d " % sv.plist[k])
        if 0==((k+1)%8):
            fd.write("\n")

    fd.close()

def proj():
    s = sieve(10000)
    not_found = True
    for p in s.plist:
        if p<1400:
            continue
        if p>=10000:
            break
        b, c = p+3330, p+6660
        if s.is_prime(b) and s.is_prime(c) and is_perm(p,b) and is_perm(p,c):
            not_found = False
            print("%d %d %d" % (p,b,c))
            print("    %d%d%d" % (p,b,c))
    if not_found:
        print("Nothing found")
    
def proj_old ( ):
    s = sieve(10000)
    #write_primes(s)
    cnt = 0
    mn_list = []
    for p in s.plist:
        if p<1000:
            continue
        if p>10000:
            break
        strp = str(p)
        if '0' in strp:
            continue
        pm = perms(p,s)
        if 3==len(pm):
            pm.sort()
            pm_str = str(pm[0]) + str(pm[1]) + str(pm[2])
            if pm_str not in mn_list:
                mn_list.append(pm_str)
    for pm in mn_list:
        x1, x2, x3 = (int(pm[:4]),int(pm[4:8]),int(pm[8:]))
        d1 = x2 - x1
        d2 = x3 - x2
        if d1==d2:
            print("    This one:")
        print("%d %d %d, %4d %4d" % (x1, x2, x3, d1, d2))

def test ( ):
    pass

if __name__=='__main__':
    proj()
    #test()

File: euler/p0049.c
#include <math.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include "primes.h"
#include "print_statements.h"

int32_t check ( pr_t x );
void proj ( void );

void test ( int32_t argc, char * argv[] );

/* ---------------------------------------------------------------- */

int32_t main ( int32_t argc, char * argv[] )
{
    //proj();
    test(argc,argv);

    return 0;
}

/* ---------------------------------------------------------------- */
void test ( int32_t argc, char * argv[] )
{
    pr_t pr = 271;
    info_print("prime_array[%llu] = %d",pr,prime_array[pr]);
    if ( is_prime(pr) )
    {
        info_print("%llu is prime",pr);
    }
    else
    {
        info_print("%llu is not prime",pr);
    }
    return;
}
/* ---------------------------------------------------------------- */

void get_dig ( pr_t * d, pr_t pr )
{
    d[0] = pr % 10;
    pr   = pr / 10;
    d[1] = pr % 10;
    pr   = pr / 10;
    d[2] = pr % 10;
    pr   = pr / 10;
    d[3] = pr % 10;
    pr   = pr / 10;
}

pr_t perm_dig ( pr_t * p, int32_t a, int32_t b, int32_t c, int32_t d )
{
    pr_t ptest = p[d] * 1000 + p[c] * 100 + p[b] * 10 + p[a];
    return ptest;
}

#define CHK(X) if ( is_prime(X) ) \
    { \
        if ( 0==cnt ) \
        { \
            fprintf(fd,"%llu -> ",pr);\
        } \
        if ( cnt<3 ) \
        { \
            plist[cnt] = X; \
        } \
        cnt++; \
        fprintf(fd," %llu ",X);\
    }


int32_t check_perms ( pr_t * plist, pr_t pr, FILE * fd )
{
    int32_t ret=0,cnt=0;
    pr_t d[4],ptest;

    get_dig(d,pr);

    ptest = perm_dig(d,3,2,1,0);
    CHK(ptest);
    ptest = perm_dig(d,3,2,0,1);
    CHK(ptest);
    ptest = perm_dig(d,3,1,2,0);
    CHK(ptest);
    ptest = perm_dig(d,3,1,0,2);
    CHK(ptest);
    ptest = perm_dig(d,3,0,2,1);
    CHK(ptest);
    ptest = perm_dig(d,3,0,1,2);
    CHK(ptest);

    ptest = perm_dig(d,2,3,1,0);
    CHK(ptest);
    ptest = perm_dig(d,2,3,0,1);
    CHK(ptest);
    ptest = perm_dig(d,2,1,3,0);
    CHK(ptest);
    ptest = perm_dig(d,2,1,0,3);
    CHK(ptest);
    ptest = perm_dig(d,2,0,3,1);
    CHK(ptest);
    ptest = perm_dig(d,2,0,1,3);
    CHK(ptest);

    ptest = perm_dig(d,1,3,2,0);
    CHK(ptest);
    ptest = perm_dig(d,1,3,0,2);
    CHK(ptest);
    ptest = perm_dig(d,1,2,3,0);
    CHK(ptest);
    ptest = perm_dig(d,1,2,0,3);
    CHK(ptest);
    ptest = perm_dig(d,1,0,3,2);
    CHK(ptest);
    ptest = perm_dig(d,1,0,2,3);
    CHK(ptest);

    ptest = perm_dig(d,0,3,2,1);
    CHK(ptest);
    ptest = perm_dig(d,0,3,1,2);
    CHK(ptest);
    ptest = perm_dig(d,0,2,3,1);
    CHK(ptest);
    ptest = perm_dig(d,0,2,1,3);
    CHK(ptest);
    ptest = perm_dig(d,0,1,3,2);
    CHK(ptest);
    ptest = perm_dig(d,0,1,2,3);
    CHK(ptest);

    if ( cnt>0 )
    {
        fprintf(fd,"\n");
    }
    if ( cnt==3 )
    {
        ret = 1;
    }

    return ret;
}

void proj ( void )
{
    pr_t lo=999, hi=10000,pr;
    int32_t cnt = 0,k,pcnt=0;
    char fname[] = "p0049_out.txt";
    FILE * fd = fopen(fname,"w");
    pr_t plist[3];

    if ( NULL==fd )
    {
        err_print("Could not open %s",fname);
        goto END;
    }

    for ( k=0; k<SMALL_PRIME_SZ && small_primes[k]<hi; ++k )
    {
        pr = small_primes[k];
        if ( pr>lo )
        {
            cnt++;
            if ( check_perms(plist,pr,fd) )
            {
                pcnt++;
                info_print("%llu -> %llu %llu %llu",
                           pr,plist[0],plist[1],plist[2]);
            }
        }
    }

    info_print("cnt = %d",cnt);
    info_print("pcnt = %d",pcnt);
    fclose(fd);
    fd = NULL;

END:
    return;
}

















